### PR TITLE
Implement #[deprecated_safe]

### PR DESCRIPTION
**Pull request by ghost**
cc #94978

### CODE REVIEW

**Comment by skippy10110** [comment]
This doesn't need a reviewer yet, I meant to create this is in the draft state.

**Comment by estebank** [comment]
@skippy10110 no problem, I'll ignore it until you ping me to let me know it is ready for review

**Comment by jhpratt** [comment]
I'm assuming the `skippy_log` stuff is just for debugging purposes?

**Comment by skippy10110** [comment]
haha, most definitely :) you should have seen the horrible things i did to the compiler before this!

**Comment by jhpratt** [comment]
Oh, I'm sure! I don't see anything that's clearly wrong in your code so far. Just the obvious todos and whatnot.

**Comment by skippy10110** [comment]
that's awesome to hear :) i was pretty much just winging it, but now i feel like i can actually justify how the code works and understand all the trait stuff (a little) better

**Comment by skippy10110** [comment]
i do have a question about reviews and my commit history

once i get things more wrapped up with a proper implementation of the attribute itself (e.g. a functioning "since" version, can only be used in the right places, etc.) and with lots of tests, proper error messages, etc.... should i rewrite the entire commit history to add each change piecemeal as if i knew what i was doing from the start?

this commit especially "THIS IS A GARBAGE COMMIT, just stashing due to data loss paranoia :)" is one i just kept force pushing over and over again as a snapshot of my whatever my current local state was

**Comment by jhpratt** [comment]
Ultimately that's up to the reviewer. Personally I'm a fan of cleaning up history, as it makes it clearer if anyone wants to come back later on to look at things.

**Comment by skippy10110** [comment]
cool, i'm a fan of tidy things so i'll make sure i do that up then

**Comment by skippy10110** [comment]
since you're following along, @jhpratt, i figured i'd ask your opinion on this (copied [from the zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/243200-t-lang.2Fmajor-changes/topic/Add.20.23.5Brustc_deprecated_safe.5D.20attribute.20to.20all.E2.80.A6.20lang-team.23147/near/276405731)):

i have a functioning ```since``` key for ```#[deprecated_safe]``` now, the semantics i chose are:
* outside of libstd ```since``` has no semantic meaning, just like ```#[deprecated]``` (applies immediately)
* within libstd, when called by 3rd party code, ```since``` will have meaning (doesn't apply until the since version)
* within libstd, when called by libstd itself, ```since``` will not have meaning (applies immediately)

so libstd can apply the attribute with e.g. a version of "TBD" without affecting 3rd party code, but it will need to update itself immediately

i felt like adding DEPRECATED_SAFE_IN_FUTURE for use within libstd was overkill, but i could do that if it's warranted

**Comment by jhpratt** [comment]
That would be in line with how `#[deprecated]` works to my understanding.

**Comment by skippy10110** [comment]
this is getting slightly ahead of the implementation work... but assuming this all lands and ```env::set_var```/```env::remove_var``` are made unsafe, what happens with ```#[test]``` code?

https://github.com/rust-lang/rust/blob/cf7ad15aa6db49609a87d54c2519116fb1cf12fb/library/std/src/process/tests.rs#L281-L294

is it possible to force individual tests to run in a single thread? i haven't found any documentation to that effect, and it's the only way that the ```unsafe {}``` wrapper around ```env::set_var```/```env::remove_var``` can be justified (afaik)

**Comment by jhpratt** [comment]
It is not currently possible to request a given test be run serially. I would support such a request for implementation, though. The rest you linked is unsound _now_; the necessary `unsafe` just demonstrates this.

And yes, definitely have the attributes for testing, but it certainly needs to be a separate PR as you've indicated.

**Comment by skippy10110** [comment]
should such a "serial testing" facility be proposed as an additional MCP? (although i'm focusing on just the ```#[deprecated_safe]``` itself first before i bite off more 😊)

although while i'm on the subject of extras, i suppose rust-analyzer is gonna need to be updated to understand the new attribute as well

**Comment by skippy10110** [comment]
oh, and i'm also wondering now if "mark ```set_var```/```remove_var``` as ```#[deprecatred_safe]```" would also needs its own dedicated MCP

**Comment by jhpratt** [comment]
I'm not too familiar with MCPs. Worth asking on Zulip for either.

**Comment by skippy10110** [comment]
kk, [i pinged my liason](https://rust-lang.zulipchat.com/#narrow/stream/243200-t-lang.2Fmajor-changes/topic/Add.20.23.5Brustc_deprecated_safe.5D.20attribute.20to.20all.E2.80.A6.20lang-team.23147/near/276428261)

**Comment by skippy10110** [comment]
(double posting [from zulip](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/.23.5Bdeprecated_safe.5D.20.2B.20Fn.2FFnMut.2FFnOnce/near/277354931) in case anyone following here happens to know the answer)

i'm getting back to this after being away for a week and have another question

one of my FIXMEs is around the fact that there are two ```confirm_fn_pointer_candidate``` functions, with one of them being related to "projection", and i'm trying to figure out if ```#[deprecated_safe]``` needs to support something here

from what i understand, projection would apply in something like this example (which doesn't compile):

```
trait Bla {
    type T;
}

impl Bla for u32 {
    type T = <std::env::set_var as FnOnce<std::ffi::OsString, std::ffi::OsString>>::Output;
}
```

is it actually possible to make something like what i've tried here that will compile? where i take a function and treat it as ```Fn```/```FnMut```/```FnOnce``` for the purposes of projection

**Comment by compiler-errors** [comment]
@bors try @rust-timer queue

(https://rust-lang.zulipchat.com/#narrow/stream/243200-t-lang.2Fmajor-changes/topic/Add.20.23.5Brustc_deprecated_safe.5D.20attribute.20to.20all.E2.80.A6.20lang-team.23147/near/277793777)

**Comment by skippy10110** [comment]
@estebank, this is ready for review to start, with a few outstanding items:
* some obvious FIXMEs around writing up new error codes with docs, etc.
* there's some stuff dumped under the stability module that might need a better home
* i have a bunch of FIXMEs with outstanding questions in the trait obligation code
    * i had to make some changes to make sure i always received trait obligations with valid cause spans to lint against when the deprecated_safe magic kicked in on unsafe fn() to closure coercions
    * this seems to relate to what's considered canonical queries, afaict
    * i'm mostly confident about the other changes, but the trait obligation/closure changes i'm pretty dubious about, linting wise...
* the ui tests still need some better organization, there's a ton of duplication due to the various combinations to test and i'm trying to make sure i have everything covered
* the [performance results](https://github.com/rust-lang/rust/pull/95025#issuecomment-1088184699) seems to show a large hit, i'm not sure if the way i'm checking for the attribute in a bunch of places is slow?

**Comment by skippy10110** [comment]
@estebank btw, the vast majority of the changes to the trait obligation code are just to make sure i get valid cause spans for linting purposes... if there's a better (higher-level?) place to lint that covers all the cases, i'd love nothing more than to rip all that out 😊

**Comment by jhpratt** [comment]
Nice work! Appears to be no difference overall.

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 1]
```diff
@@ -2021,7 +2024,13 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
                         let fn_sig = self.normalize(fn_sig, location);
 
-                        let ty_fn_ptr_from = tcx.safe_to_unsafe_fn_ty(fn_sig);
+                        let ty_fn_ptr_from = match ptr_cast {
+                            PointerCast::UnsafeFnPointer => tcx.safe_to_unsafe_fn_ty(fn_sig),
+                            PointerCast::DeprecatedSafeFnPointer => {
+                                tcx.unsafe_to_safe_fn_ty(fn_sig)
+                            }
+                            _ => unreachable!(),
+                        };
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/type_check/mod.rs:2033]: why is that needed?

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs [hunk 2]
```diff
@@ -216,6 +216,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         // This is a no-op at the LLVM level.
                         operand.val
                     }
+                    mir::CastKind::Pointer(PointerCast::DeprecatedSafeFnPointer) => {
+                        // This is a no-op at the LLVM level.
+                        operand.val
+                    }
```
Comment (lcnr) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:222]: could merge this with the branch above instead of copying it

File: compiler/rustc_const_eval/src/interpret/cast.rs [hunk 3]
```diff
@@ -74,6 +74,17 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                let src = self.read_immediate(src)?;
+                match cast_ty.kind() {
+                    ty::FnPtr(_) => {
+                        // No change to value
+                        self.write_immediate(*src, dest)?;
+                    }
+                    _ => span_bug!(self.cur_span(), "unsafe fn to fn cast on {:?}", cast_ty),
+                }
+            }
+
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/interpret/cast.rs:87]: same here

**Review by lcnr — COMMENTED** [review]
I've long wanted to look into adding a lint if certain predicates are used, mostly to deprecate the `Eq` and `PartialEq` impls for function pointers.

While the approach of linting when not in canonical queries works right now, my goal - and of @rust-lang/wg-traits afaik -  is to slowly work towards always using queries for trait solving. At this point your code would break.

My expected solution here is to add "this should lint" to the output of a successful selection then let the caller emit the error. I didn't yet look into what's the best way to do this, but without too much thought it might make sense to just change the output of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/struct.SelectionContext.html#method.confirm_candidate to have a `lint_on_use: Option<SelectionLint>` field. We would then have to propagate this upwards to outside of the query.

Probably makes sense for myself to look into this more, now that someone else also wants that. I will only get to that at some point after the 25th April though.

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 1 above)
Comment (skippy10110) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: i have:
```
CastKind::Pointer(
    ptr_cast @ (PointerCast::UnsafeFnPointer
    | PointerCast::DeprecatedSafeFnPointer),
) => {
```
followed by:
let ty_fn_ptr_from = match ptr_cast {
    PointerCast::UnsafeFnPointer => ...,
    PointerCast::DeprecatedSafeFnPointer => ...,
    _ => unreachable!(),
};
```

being a rust noob, i'm kinda hoping there's a better way here 😊 i want to match on the two variants that i've already guaranteed are the only two i can see

Comment (lcnr) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: ah, the branches are actually different :sweat_smile: `unsafe -> safe` vs `safe -> unsafe` xx

sry

**Comment by skippy10110** [comment]
i was thinking i need a "attach lint for later reporting higher up" facility when selection is successful, since all the the nice error reporting seems to come from returning a SelectionError instead of a Selection, so that's perfect to hear that would be a good direction to go!

i wasn't sure if that would be too obtrusive a change, but it sounds like my hacky don't query change is going the opposite direction of what we want, based on what you're saying

i'll start taking a look at this and see if i can make any progress on it

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 1 above)
Comment (skippy10110) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: i'll run into the same thing for determining the message when i merge [these two arms](https://github.com/rust-lang/rust/blob/5c1f53ea37387c615c573c7a9a01822091edc41e/compiler/rustc_const_eval/src/interpret/cast.rs#L66-L86) like you suggested as well

Comment (skippy10110) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: no worries 😊 do you know if the ```unreachable!()``` is possible to remove though? it bothers me 🤓

Comment (lcnr) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: i don't think it can easily be removed :sweat_smile: you could merge `PointerCast::DeprecatedSafeFnPointer` and `PointerCast::UnsafeFnPointer` into
`PointerCast::FnPointer(FnCastKind)` but I don't think that actually improves anything :sweat_smile:

Comment (skippy10110) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2033]: thanks! guess i'll have to live with it 😊 i'll still go ahead and merge the arms in the other two places though

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs (hunk 2 above)
Comment (skippy10110) [reply compiler/rustc_codegen_ssa/src/mir/rvalue.rs:220]: done

File: compiler/rustc_const_eval/src/interpret/cast.rs (hunk 3 above)
Comment (skippy10110) [reply compiler/rustc_const_eval/src/interpret/cast.rs:89]: done

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 1 above)
Comment (skippy10110) [reply compiler/rustc_borrowck/src/type_check/mod.rs:2034]: the unreachable!() is commented now

**Comment by skippy10110** [comment]
@lcnr, @estebank, i've made some progress on the obligation linting, which affects the "implement deprecated_safe for fn() pointer closure coercions" commit in this PR

assuming we go the route of putting a new linting system in place that this PR can then use, the commit i mentioned would change from:
* https://github.com/skippy10110/rust/commit/10c17928e8a01bcaf4e02303d09d4d23cbccdb6a
to something like:
* https://github.com/skippy10110/rust/commit/1a2ea4716ebae7c40ab58d8c738753cbc9c3a642

the rest of the commits in this PR are unaffected by this

**Comment by JohnCSimon** [comment]
Ping from triage:
@skippy10110 - can you please address the conflicts and build failures and post your status?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by wesleywiser** [comment]
Visited during compiler team triage meeting. Looks like this is blocked on #96092. Marking as such for now.

**Comment by jhpratt** [comment]
As this PR's author has deleted their account, I have backed up the state of this pull request on my fork. [It is located here](https://github.com/jhpratt/rust/tree/implement-deprecated-safe). This should make it easier to use existing code if anyone would like to take it over.
